modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeReplaceRemoveRaceTest.java

```diff
@@ -0,0 +1,421 @@
+
+        lockTrackerManager.start();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        if (pageMem != null)
+            pageMem.stop(true);
+
+        if (lockTrackerManager != null)
+            lockTrackerManager.stop();
+    }
+
+    /**
+     * @return Page memory.
+     */
+    protected PageMemory createPageMemory() {
+        DataRegionConfiguration plcCfg = new DataRegionConfiguration()
+            .setInitialSize(1024 * MB)
+            .setMaxSize(1024 * MB);
+
+        PageMemory pageMem = new PageMemoryNoStoreImpl(log,
+            new UnsafeMemoryProvider(log),
+            PAGE_SIZE,
+            plcCfg,
+            new DataRegionMetricsImpl(plcCfg, new GridTestKernalContext(log())),
+            true);
+
+        pageMem.start();
+
+        return pageMem;
+    }
+
+    /**
+     * @return Allocated meta page ID.
+     * @throws IgniteCheckedException If failed.
+     */
+    private FullPageId allocateMetaPage() throws IgniteCheckedException {
+        return new FullPageId(pageMem.allocatePage(CACHE_ID, PageIdAllocator.INDEX_PARTITION, PageIdAllocator.FLAG_IDX), CACHE_ID);
+    }
+
+    /**
+     * Short for {@code T2<Integer, Integer>}.
+     */
+    protected static class Pair extends T2<Integer, Integer> {
+        /** Serial version uid. */
+        private static final long serialVersionUID = 0L;
+
+        /**
+         * @param key Key.
+         * @param val Value.
+         */
+        public Pair(Integer key, Integer val) {
+            super(key, val);
+        }
+
+        /**
+         * No-arg constructor for {@link Externalizable}.
+         */
+        public Pair() {
+        }
+    }
+
+    /**
+     * Test tree, maps {@link Integer} to {@code Integer}.
+     */
+    protected static class TestPairTree extends BPlusTree<Pair, Pair> {
+        /**
+         * @param reuseList Reuse list.
+         * @param canGetRow Can get row from inner page.
+         * @param cacheId Cache ID.
+         * @param pageMem Page memory.
+         * @param metaPageId Meta page ID.
+         * @throws IgniteCheckedException If failed.
+         */
+        public TestPairTree(
+            ReuseList reuseList,
+            boolean canGetRow,
+            int cacheId,
+            PageMemory pageMem,
+            long metaPageId,
+            PageLockTrackerManager lockTrackerManager
+        ) throws IgniteCheckedException {
+            super(
+                "test",
+                cacheId,
+                null,
+                pageMem,
+                null,
+                new AtomicLong(),
+                metaPageId,
+                reuseList,
+                new IOVersions<>(new TestPairInnerIO()),
+                new IOVersions<>(new TestPairLeafIO()),
+                PageIdAllocator.FLAG_IDX,
+                new FailureProcessor(new GridTestKernalContext(log)) {
+                    @Override public boolean process(FailureContext failureCtx) {
```

Comment:
/** {@inheritDoc} */

Reply:
Do we write it in inner classes? It's not like I care, just wandering)